### PR TITLE
Feature/gRPC

### PR DESCRIPTION
**Pull request by the-man-w-laughs**
Change http to gRPC in gathering FullProjectInfo form TeamHub for ReportHub.

Tech-stack:
- code first approach to generate gRPC service and client;

### CODE REVIEW

File: src/Shared/Extensions/ClaimsPrincipalExtensions.cs [hunk 1]
```diff
@@ -24,6 +25,7 @@ public static int GetUserId(this ClaimsPrincipal principal)
         }
 
         var claim = principal.FindFirst(x => x.Type.Equals(ClaimTypes.NameIdentifier));
+
         if (claim == null)
         {
             throw new InvalidOperationException("The 'sub' claim is not present in the principal.");
```
Comment (mebr1) [inline src/Shared/Extensions/ClaimsPrincipalExtensions.cs:31]: It's not necessary at the moment, but it's better to move exception messages to a separate file

File: src/services/ReportHub/ReportHub.BLL/Extensions/FullProjectParserExtensions.cs [hunk 2]
```diff
@@ -1,23 +1,26 @@
...
+using Shared.gRPC.FullProjectResponse;
 
 namespace ReportHub.BLL.Extensions
 {
     public static class FullProjectParserExtensions
     {
-        public static string ToReport(this FullProjectResponseDto project)
+        public static string ToReport(this FullProjectInfoResponse project)
         {
             var sb = new StringBuilder();
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Extensions/FullProjectParserExtensions.cs]: Let's make a more specific variable name

File: src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs [hunk 3]
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
             //     return latestReport;
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs]: Remove the commented code

File: src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs [hunk 4]
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
...
             // }
 
-            FullProjectResponseDto fullProjectResponseDto = await GetFullProjectDataAsync(
-                projectId
-            );
+            var request = new FullProjectInfoRequest() { ProjectId = projectId, UserId = userId };
+
+            FullProjectInfoResponse fullProjectResponseDto =
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs]: Use var, don't explicitly specify the type

File: src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs [hunk 5]
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
...
+                await _fullProjectInfoService.GetFullProjectInfoAsync(request);
 
             if (fullProjectResponseDto == null)
             {
                 throw new NotFoundException(
                     $"Cannot fetch full project data for with id {projectId}."
                 );
             }
+
             string reportContent = fullProjectResponseDto.ToReport();
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs]: Same. Check throughout the code for explicit type annotations

File: src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    public static class GrpcRegistrationExtensions
+    {
+        public static void ReristerRrpcService(this IServiceCollection services)
+        {
+            services.AddCodeFirstGrpc();
+        }
+
+        public static WebApplicationBuilder ConfigureWebHost(this WebApplicationBuilder builder)
+        {
+            builder.WebHost.ConfigureKestrel(options =>
```
Comment (mebr1) [inline src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs]: Move the Kestrel configuration to the appsettings file

File: src/services/TeamHub/TeamHub.DAL/Extensions/DalRegistrationExtensions.cs [hunk 7]
```diff
@@ -17,8 +17,10 @@ IConfiguration configuration
             services.AddDbContext<TeamHubDbContext>(builder =>
             {
                 var connectionString = configuration.GetConnectionString("Default");
+
                 if (string.IsNullOrWhiteSpace(connectionString))
```
Comment (mebr1) [inline src/services/TeamHub/TeamHub.DAL/Extensions/DalRegistrationExtensions.cs:21]: Add curly braces for if

File: src/services/ReportHub/ReportHub.BLL/Extensions/GrpcRegistrationExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace ReportHub.BLL.Extensions
+{
+    public static class GrpcRegistrationExtensions
+    {
+        public static void RegisterGrpcClient(
+            this IServiceCollection services,
+            IConfiguration config
+        )
+        {
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Extensions/GrpcRegistrationExtensions.cs:16]: Let's use AddGrpcClient for configuration

File: src/services/ReportHub/ReportHub.BLL/Extensions/Utils/GrpcClientFactory.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
+using Grpc.Net.Client;
+using ProtoBuf.Grpc.Client;
+
+namespace ReportHub.BLL.Extensions.Utils
+{
+    public class GrpcClientFactory
```
Comment (mebr1) [inline src/services/ReportHub/ReportHub.BLL/Extensions/Utils/GrpcClientFactory.cs]: Where are you using this class? I didn't find any references to it

File: src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                    }
+                );
+            });
+
+            return builder;
+        }
+
+        public static void UseGrpcService(this IApplicationBuilder app)
+        {
+            app.UseRouting();
```
Comment (mebr1) [inline src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs]: UseRouting should not be placed here

**Review by mebr1 — CHANGES_REQUESTED** [review]

File: src/Shared/Extensions/ClaimsPrincipalExtensions.cs (hunk 1 above)
Comment (the-man-w-laughs) [reply src/Shared/Extensions/ClaimsPrincipalExtensions.cs:31]: What if I need to throw an exceptions with interpolated string messages or create a class only for parameterless messages?

File: src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs (hunk 6 above)
Comment (the-man-w-laughs) [reply src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs]: Cannot find any info how to do that.

File: src/services/ReportHub/ReportHub.BLL/Extensions/GrpcRegistrationExtensions.cs (hunk 8 above)
Comment (the-man-w-laughs) [reply src/services/ReportHub/ReportHub.BLL/Extensions/GrpcRegistrationExtensions.cs:16]: I use code first approach to configure gRPC and no client code is generated.
I didn't find any overloads or info how to inject client using AddGrpcClient if i don't have a client class.

File: src/services/ReportHub/ReportHub.BLL/Extensions/Utils/GrpcClientFactory.cs (hunk 9 above)
Comment (the-man-w-laughs) [reply src/services/ReportHub/ReportHub.BLL/Extensions/Utils/GrpcClientFactory.cs]: I removed it

File: src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs (hunk 6 above)
Comment (mebr1) [reply src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs]: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0

File: src/Shared/Extensions/ClaimsPrincipalExtensions.cs (hunk 1 above)
Comment (mebr1) [reply src/Shared/Extensions/ClaimsPrincipalExtensions.cs:31]: You can create custom exceptions and overload their constructors with various parameters. Within the service, you simply call the required exception constructor, eliminating the need to explicitly write the message. For constant messages, you can create a separate class and directly insert them into the exception

File: src/services/ReportHub/ReportHub.BLL/Extensions/FullProjectParserExtensions.cs (hunk 2 above)
Comment (mebr1) [reply src/services/ReportHub/ReportHub.BLL/Extensions/FullProjectParserExtensions.cs]: I mean that the variable name "sb" isn't informative. While you work with it extensively, it's difficult to determine its contents from the name alone

**Review by mebr1 — CHANGES_REQUESTED** [review]

**Review by mebr1 — APPROVED** [review]

**Pull request merged**